types/validator_set.go

```diff
@@ -78,44 +78,52 @@ func (vals *ValidatorSet) IncrementProposerPriority(times int) {
 		panic("Cannot call IncrementProposerPriority with non-positive times")
 	}
 
-	const shiftEveryNthIter = 10
+	// cap the difference between priorities to be proportional to 2*totalPower by
+	// re-normalizing priorities, i.e., rescale all priorities by multiplying with:
+	//  2*totalVotingPower/(maxPriority - minPriority)
+	diff := computeMaxMinPriorityDiff(vals)
+	threshold := 2 * vals.TotalVotingPower()
+	// the 2nd check (threshold > 0) is merely a sanity check which could be
+	// removed if all tests would init. voting power appropriately;
+	// i.e. threshold should always be > 0
+	if diff > threshold && threshold > 0 {
+		// div = Ceil((maxPriority - minPriority) / 2*totalVotingPower)
+		// threshold > 0 and diff > threshold guarantees (diff / threshold > 0):
+		div := int64(math.Ceil(float64(diff) / float64(threshold)))
+		vals.dividePrioritiesBy(div)
+	}
+
 	var proposer *Validator
 	// call IncrementProposerPriority(1) times times:
 	for i := 0; i < times; i++ {
-		shiftByAvgProposerPriority := i%shiftEveryNthIter == 0
-		proposer = vals.incrementProposerPriority(shiftByAvgProposerPriority)
-	}
-	isShiftedAvgOnLastIter := (times-1)%shiftEveryNthIter == 0
-	if !isShiftedAvgOnLastIter {
-		validatorsHeap := cmn.NewHeap()
-		vals.shiftByAvgProposerPriority(validatorsHeap)
+		proposer = vals.incrementProposerPriority()
 	}
+	vals.shiftByAvgProposerPriority()
```

Comment:
Are we sure averaging only once per call is safe? If we don't average every `n` times, and `times` is really large (maybe when catching up syncing a node), it seems like we might run into the clipping logic and end up with a different final set of priorities (and possibly validators).

Reply:
Good point. Given the speed tendermint creates blocks, being offline for a few weeks (or months) may cause the clipping to kick in. Not sure how realistic that actually is but we should probably account for this.  Previously, we had every 10th iter. Maybe we should just make this number a bit larger then?